src/ProxyClient/AbstractProxyClient.php

```diff
@@ -120,189 +69,46 @@ protected function getBaseUrl()
+            $collection = new ExceptionCollection();
+            foreach ($e->getExceptions() as $exception) {
+                // A workaround for php-http currently lacking differentiation
+                // between client, server and networking errors.
+                if (!$exception->getResponse()) {
+                    // Assume networking error if no response was returned.
+                    $collection->add(
+                        ProxyUnreachableException::proxyUnreachable($exception)
+                    );
+                } else {
+                    $collection->add(
+                        ProxyResponseException::proxyResponse($exception)
+                    );
+                }
             }
 
-            $collection->add($e);
+            throw $collection;
         }
-
-        throw $collection;
+        
+        return count($queue);
     }
-
-    /**
-     * Filter a URL
-     *
-     * Prefix the URL with "http://" if it has no scheme, then check the URL
-     * for validity. You can specify what parts of the URL are allowed.
-     *
-     * @param string   $url
-     * @param string[] $allowedParts Array of allowed URL parts (optional)
-     *
-     * @throws InvalidUrlException If URL is invalid, the scheme is not http or
-     *                             contains parts that are not expected.
-     *
-     * @return string The URL (with default scheme if there was no scheme)
-     */
-    protected function filterUrl($url, array $allowedParts = array())
+    
+    protected function queueRequest($method, $url, array $headers = array())
     {
-        // parse_url doesn’t work properly when no scheme is supplied, so
-        // prefix URL with HTTP scheme if necessary.
-        if (false === strpos($url, '://')) {
-            $url = sprintf('%s://%s', $this->getDefaultScheme(), $url);
-        }
-
-        if (!$parts = parse_url($url)) {
-            throw InvalidUrlException::invalidUrl($url);
-        }
-        if (empty($parts['scheme'])) {
-            throw InvalidUrlException::invalidUrl($url, 'empty scheme');
-        }
-
-        if (!in_array(strtolower($parts['scheme']), $this->getAllowedSchemes())) {
-            throw InvalidUrlException::invalidUrlScheme($url, $parts['scheme'], $this->getAllowedSchemes());
-        }
-
-        if (count($allowedParts) > 0) {
-            $diff = array_diff(array_keys($parts), $allowedParts);
-            if (count($diff) > 0) {
-                throw InvalidUrlException::invalidUrlParts($url, $allowedParts);
-            }
-        }
-
-        return $url;
+        $this->queue->add(new InvalidationRequest($method, $url, $headers));
     }
-
-    /**
-     * Get default scheme
-     *
-     * @return string
-     */
-    protected function getDefaultScheme()
+    
+    protected function initQueue(array $servers, $baseUrl)
     {
-        return 'http';
+        $this->queue = new RequestQueue($servers, $baseUrl);
```

Comment:
No objections to constructing the queue object here? It would be better (in terms of DI) to pass one to the AbstractProxyClient’s constructor, but I think that may be confusing for users.